plugins/scaffolder-react/src/next/components/Stepper/Stepper.test.tsx

```diff
@@ -436,4 +436,48 @@ describe('Stepper', () => {
       expect(getByRole('textbox', { name: 'field1' })).toBeInTheDocument();
     });
   });
+
+  it('should call validator in extension if no key initially exists in formData', async () => {
+    const manifest: TemplateParameterSchema = {
+      title: 'Fill in some steps',
+      steps: [
+        {
+          title: 'Step 1',
```

Comment:
I feel like I'm missing a piece here.

If the thing is marked as required, and it's empty, does the `validator` run or does the jsonchema validation error first?

I'm feeling that in any case the validator will sometimes be given an empty value, and then it needs to know how to deal with that right? Is this then a breaking change that all validators need to know that they might get given an empty value, and in that case we need to guard before doing anything else in the `validator`?

Reply:
The way it currently works is that `handleNext` is assigned to the `onSubmit` of the `@rjsf` form.

When next is clicked, the `@rjsf` required validations are run, and only when all the required validations have been cleared by the user does the code reach this line in `handleNext`:

```ts
const returnedValidation = await validation(formData);
```

The validate function returned from `createAsyncValidators` works by iterating over the keys of `formData`. 

In the example in the test in my PR:

```ts
const manifest: TemplateParameterSchema = {
  title: 'Fill in some steps',
  steps: [
    {
      title: 'Step 1',
      schema: {
        type: 'object',
        properties: {
          name: {
            type: 'string',
            'ui:field': 'Mock',
          },
        },
      },
    },
  ],
};
```

If there is no `name` field in the `formData`, then no validations will be run.  If `name` was required there is a chance that it make it into the formData if the user clicked `next` without filling out the field, but this is not guaranteed.

In order for this too work properly, the formData needs to be initialised properly with undefined values which is what `getDefaultFormState` does.

You cannot guarantee that the custom extension will be executed otherwise.

Does this help?

Reply:
> If the thing is marked as required, and it's empty, does the `validator` run or does the jsonchema validation error first and not run it?
> 

@benjdlambert sorry, I re-read your question.

If the field is required and there is no key in `formData` then the `validator` will not run.

I would expect it to run whether it is required or not and I think this is the way it currently works in `@rjsf@3.2.0`.

I think this is a better answer than above :).

Reply:
Ok, so I just checked the current stable scaffolder, and you're right, that it does indeed call the validation. Which actually throws an error for things like the `LowerCaseFieldExtension` that we have in `packages/app` as the field is not required and calls it with an undefined value as the value. 

https://github.com/backstage/backstage/blob/master/packages/app/src/components/scaffolder/customScaffolderExtensions.tsx#L60-L64

My initial thought is that this behaviour is wrong, so I want to understand what's the use case for running this validation function when there is no value provided in the form and it's not required? Maybe it makes sense, in which case we should enforce that the first argument should be `string?:` instead.

Reply:
@benjdlambert I think the point you raise is a good one.  The reasons I raised this is because:

1.  It is different that what is currently happening in the `<ScaffolderPage />`
2.  I actually thought there was a bug, but it turns out that the custom validations do not get run until you have cleared any required field validations and then clicked next again.

Our users originally reported this after upgrading to `<NextScaffolderPage />` because the behaviour is different but I agree that the behaviour is correct.

Reply:
> I actually thought there was a bug, but it turns out that the custom validations do not get run until you have cleared any required field validations and then clicked next again.

Yeh this is interesting - I guess this is because the synchronous `jsonschema` validations run first, and then we call our validation functions. I guess we could move out that validation to do them all at the same time, so we turn off `validate` in the form, and validate it separately, but maybe not worth it.

> Our users originally reported this after upgrading to <NextScaffolderPage /> because the behaviour is different but I agree that the behaviour is correct.

This is the reason I'm asking for a use case why you would ever want to do this, I really feel like it was potentially a bug before, and that you shouldn't need this, and I'm feeling that we just add a special callout for this as a potentially breaking change in https://backstage.io/docs/features/software-templates/testing-scaffolder-alpha#breaking-changes.

Feels like if we were dependant on that code to run in the validation, I don't see what value it can be doing for the validation part when there's no value and it's not required as the `jsonschema` deals with that. But if we need to be running some additional code then maybe it's a separate API for that, whatever that looks like.

Does this make sense?

Reply:
sorry I mean the behaviour in the `NextScaffolderPage` is correct and this PR should be closed.

The old behviour is wrong.  Validations should not be ran on optional fields.

Reply:
@dagda1 sorry yeah I think we're saying the same thing and we're agreed. Given that this wont be called anymore, do you want to repurpose this PR or maybe create a separate on for adding a bit the the breaking changes section as just an additional callout that if you were dependant on this it's been changed or do you not think it's worth it?

Reply:
@benjdlambert i think it might be confusing to mention the wrong behaviour.

Validations should not run on optional fields.  That makes sense

If enough people mention it in discord then something can be added

Reply:
Ok - makes sense. Closing this PR then. Thanks for the patience! :pray: